### ISSUE TITLE
Bokeh issue with runipy

### ISSUE THREAD

**Issue by tjimation** [bokeh/bokeh#4143]
For Bokeh .9 and .10, we were able to run a notebook (using bokeh) thru runipy and generate a html page.  However, this is broken with bokeh .11*.

So in .11*, we are getting an error message:
```
Traceback (most recent call last):
  File ........./notebook/runner.py, line 37, in run
    r.run_notebook()
  File ........................../lib/python2.7/site-packages/runipy/notebook_runner.py, line 206, in run_notebook
    self.run_cell(cell)
  File ........................../lib/python2.7/site-packages/runipy/notebook_runner.py, line 178, in run_cell
    raise NotImplementedError('unhandled iopub message: %s' % msg_type)
NotImplementedError: unhandled iopub message: comm_open
```
So somehow, a message of "comm_open" (different than .9 and .10 behavior) is passed back to runipy/notebook_runner.py which can't handle it.

Thru google, I found a link of a custom notebook_runner.py which seems altering the code for other purpose, but it introduced a block of code to handle such (unfortunately this doesn't work for my purpose).
http://www.xavierdupre.fr/app/pyquickhelper/helpsphinx/_modules/pyquickhelper/ipythonhelper/notebook_runner.html

            elif msg_type == 'comm_open' or msg_type == 'comm_msg':
                # widgets in a notebook
                out.data = content["data"]
                out.comm_id = content["comm_id"]

This block of code is not in the official version of notebook_runner.py.
https://github.com/paulgb/runipy/blob/master/runipy/notebook_runner.py

So my question is why such message of "comm_open" is introduced in .11*.  How can I get around this problem?  BTW, this is my first post to github (and I am new to bokeh/runipy), so if I addressed this question at the wrong forum or in a wrong way, please direct me to the proper channel.

Thanks.

**Comment by bryevdv** [comment]
"Notebook Comms" are what implement `push_notebook`, which allows, e.g. to have streaming data updating plots in a running Jupyter notebook. It works with the standard Jupyter Notebook, which is all we test against. This is a critical and important feature for Jupyter notebooks, so it cannot be removed. A solution probably involves two things:

* In Bokeh, update this block to fail gracefully (i.e., put in a 'try'/'catch') https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/embed.coffee#L24-L28

* make `notebook/runner.py` not fall over when it gets comms messages. Ideally it would return some kind of appropriate error condition or exception (I don't know enough about Jupyter internals to say for sure) but at a minimum, don't explode. 

So, basically I don't think there is any solution that does not involve changes both in Bokeh and upstream. There is nothing we can do to not-set the comms request, which means `runipy` has to change to handle it more gracefully.

**Comment by bryevdv** [comment]
ping @damianavila

**Comment by bryevdv** [comment]
It's possible we could add an environment variable to control the use of comms. But that involves non-trivial plumbing to communicate from the python side to the JS side. I'm generally not positive on the idea to being with, and I definitely don't regard it as a valuable use of our very limited resources, but a pull request would certainly be considered.

**Comment by bryevdv** [comment]
I have done the first bullet point above in #4144 I you have the ability to test a build based off that PR, with some changes to `notebook/runner.py` as I described, that would let us know if this solution could work for both use cases.

**Comment by bryevdv** [comment]
@tjimation I have pushed the PR for our end of the work. It will be available in a dev build tomorrow, and in `0.12`. I am going to close this issue as `upstream` for now (given the work that has to happen in the `runipy` side) unless there is new information to consider, in which case please-reopen.

**Comment by tjimation** [comment]
I tested the try /catch change described in 4144.  I agree with you that it makes sense to have it anyway to address other issues, but it doesn't help my issue.  

Below is a block of code in runipy/notebook_runner.py (yes, I understand this is not bokeh code, and you labeled "upstream", but) I want to point out:

A. Bokeh now returns msg_type of "comm_open" and "comm_msg" that runipy/notebook_runner.py can't handle.  Per your explanation above, that has to be done to "support stream data updating plot".  My question is, can we get an option in bokeh to completely disable such option (since I assume many people are not using this) and so such message don't have to send over to runipy (upstream)?

B. Even if I added such new msg_type to this line
elif msg_type in ('display_data', 'pyout', "comm_open", "comm_msg"):
I am getting raw js text back now than a plot before (bokeh 9 and 10).  So are there any changes to the ouptut (format or the tags wrapping around it)?  I am still researching this on my end (to compare the difference of the output)

Again, I understand it would be nice that runipy (upstream) can handle it + nice to have new feature, but I am just wondering if more (option) can be done at Bokeh side (A) + trying to understand if there is any other changes causing a different behavior than older version of Bokeh (B).

Thanks again for your replies and time.


    def run_cell(self, cell):
        '''
        Run a notebook cell and update the output of that cell in-place.
        '''
        logging.info('Running cell:\n%s\n', cell.input)
        self.kc.execute(cell.input)
        reply = self.kc.get_shell_msg()
        status = reply['content']['status']
        if status == 'error':
            traceback_text = 'Cell raised uncaught exception: \n' + \
                '\n'.join(reply['content']['traceback'])
            logging.info(traceback_text)
        else:
            logging.info('Cell returned')

        outs = list()
        while True:
            try:
                msg = self.kc.get_iopub_msg(timeout=1)
                if msg['msg_type'] == 'status':
                    if msg['content']['execution_state'] == 'idle':
                        break
            except Empty:
                # execution state should return to idle before the queue becomes empty,
                # if it doesn't, something bad has happened
                raise

            content = msg['content']
            msg_type = msg['msg_type']

            # IPython 3.0.0-dev writes pyerr/pyout in the notebook format but uses
            # error/execute_result in the message spec. This does the translation
            # needed for tests to pass with IPython 3.0.0-dev
            notebook3_format_conversions = {
                'error': 'pyerr',
                'execute_result': 'pyout'
            }
            msg_type = notebook3_format_conversions.get(msg_type, msg_type)

            out = NotebookNode(output_type=msg_type)

            if 'execution_count' in content:
                cell['prompt_number'] = content['execution_count']
                out.prompt_number = content['execution_count']

            if msg_type in ('status', 'pyin', 'execute_input'):
                continue
            elif msg_type == 'stream':
                out.stream = content['name']
                # in msgspec 5, this is name, text
                # in msgspec 4, this is name, data
                if 'text' in content:
                    out.text = content['text']
                else:
                    out.text = content['data']
                #print(out.text, end='')
            elif msg_type in ('display_data', 'pyout'):
                for mime, data in content['data'].items():
                    try:
                        attr = self.MIME_MAP[mime]
                    except KeyError:
                        raise NotImplementedError('unhandled mime type: %s' % mime)

                    setattr(out, attr, data)
                #print(data, end='')
            elif msg_type == 'pyerr':
                out.ename = content['ename']
                out.evalue = content['evalue']
                out.traceback = content['traceback']

                #logging.error('\n'.join(content['traceback']))
            elif msg_type == 'clear_output':
                outs = list()
                continue
            else:
                raise NotImplementedError('unhandled iopub message: %s' % msg_type)
            outs.append(out)
        cell['outputs'] = outs

        if status == 'error':
            raise NotebookError(traceback_text)

**Comment by bryevdv** [comment]
HI @tjimation To be clear, I'm not suggesting you "handle" the messages except as far as to not fall over (e.g. not raise an exception) when they arrive. You can drop them on the floor or ignore them, I don't think it matters. 

As I mentioned before, the team is currently very overextended. `push_notebook` is an important feature for jupyter notebooks, and jupypter notebooks are all we can afford to prioritize. "adding an option" to disable actually would involve some fairly complicated cross-language / cross-runtime plumbing, with all the tests and documentation and maintenance commitments that entails. I can't justify asking anyone on the core team to spend time on this. Certainly if you wanted to submit a PR for such a feature, it would be considered. 

I should also apprise you that there is a possibility in the future that *all* notebook rendering will go through comms, in order reduce the number of code paths we have to maintain (we can re-use the Bokeh server protocol over notebook comms instead of doing things completely differently in the notebook as we currently do)
